src/server/gc_worker/compaction_runner.rs

```diff
@@ -0,0 +1,728 @@
+                num_total_entries,
+                num_rows,
+                region_id,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Estimates the number of discardable MVCC entries based on GC safe point
+    fn get_estimated_discardable_entries(
+        &self,
+        num_entries: u64,
+        oldest_ts: TimeStamp,
+        newest_ts: TimeStamp,
+        gc_safe_point: u64,
+    ) -> u64 {
+        if num_entries == 0 || oldest_ts >= newest_ts {
+            return 0;
+        }
+        let oldest_ts = oldest_ts.into_inner();
+        let newest_ts = newest_ts.into_inner();
+
+        if gc_safe_point >= newest_ts {
+            return num_entries;
+        }
+        if gc_safe_point < oldest_ts {
+            return 0;
+        }
+
+        let total_range = newest_ts - oldest_ts;
+        let discardable_range = gc_safe_point - oldest_ts;
+        let portion = (discardable_range as f64) / (total_range as f64);
+        (num_entries as f64 * portion).round() as u64
+    }
+
+    /// Calculates compaction score based on tombstones and discardable entries
+    fn get_compact_score(
+        &self,
+        num_tombstones: u64,
+        num_discardable: u64,
+        num_total_entries: u64,
+        config: &GcConfig,
+    ) -> f64 {
+        if num_total_entries == 0 || num_total_entries < num_discardable {
+            return 0.0;
+        }
+
+        if !config.enable_compaction_filter {
+            // Only consider deletes (tombstones)
+            let ratio = num_tombstones as f64 / num_total_entries as f64;
+            if num_tombstones < config.auto_compaction.tombstones_num_threshold
+                && ratio < config.auto_compaction.tombstones_percent_threshold as f64 / 100.0
+            {
+                return 0.0;
+            }
+            return num_tombstones as f64 * ratio;
+        }
+
+        // When compaction filter is enabled, ignore tombstone threshold,
+        // just add deletes to redundant keys for scoring.
+        let ratio = (num_tombstones + num_discardable) as f64 / num_total_entries as f64;
+        if num_discardable < config.auto_compaction.redundant_rows_threshold
+            && ratio < config.auto_compaction.redundant_rows_percent_threshold as f64 / 100.0
+        {
+            return 0.0;
+        }
+        num_discardable as f64 * ratio
+    }
+
+    fn sleep_or_stop(&mut self, timeout: Duration) -> bool {
+        if self.is_stopped {
+            return true; // Already stopped
+        }
+        match self.stop_signal_receiver.as_ref() {
+            Some(rx) => match rx.recv_timeout(timeout) {
+                Ok(_) => {
+                    self.is_stopped = true;
+                    true // Stop requested
+                }
+                Err(mpsc::RecvTimeoutError::Timeout) => false, // Continue
+                Err(mpsc::RecvTimeoutError::Disconnected) => {
+                    panic!("stop_signal_receiver unexpectedly disconnected")
```

Comment:
Better not panic, I think add a error log and return is enough.

Reply:
Done